src/bmp_matcher.rs

```diff
@@ -0,0 +1,305 @@
+	pub fn get_index(&self) -> Option<usize>
+	{
+		self.index
+	}
+
+	/// Get any serial number previously set with `.serial()`.
+	pub fn get_serial(&self) -> Option<&str>
+	{
+		self.serial.as_deref()
+	}
+
+	/// Get any port path previously set with `.port()`.
+	pub fn get_port(&self) -> Option<PortId>
+	{
+		self.port.clone()
+	}
+
+	/// Find all connected Black Magic Probe devices that match from the command-line criteria.
+	///
+	/// This uses the `serial_number`, `index`, and `port` values from `matches`, treating any that
+	/// were not provided as always matching.
+	///
+	/// This function returns all found devices and all errors that occurred during the search.
+	/// This is so errors are not hidden, but also do not prevent matching devices from being found.
+	/// However, if the length of the error `Vec` is not 0, you should consider the results
+	/// potentially incomplete.
+	///
+	/// The `index` matcher *includes* devices that errored when attempting to match them.
+	pub fn find_matching_probes(&self) -> BmpMatchResults
+	{
+		let mut results = BmpMatchResults {
+			found: Vec::new(),
+			filtered_out: Vec::new(),
+			errors: Vec::new(),
+		};
+
+		let devices = match list_devices() {
+			Ok(d) => d,
+			Err(e) => {
+				results.errors.push(e.into());
+				return results;
+			},
+		};
+
+		// Filter out devices that don't match the Black Magic Probe's vid/pid in the first place.
+		let devices = devices.filter(|dev| {
+			let vid = dev.vendor_id();
+			let pid = dev.product_id();
+			BmpPlatform::from_vid_pid(Vid(vid), Pid(pid)).is_some()
+		});
+
+		devices
+			.enumerate()
+			.map(|(index, device_info)| self.matching_probe(index, device_info))
+			.collect()
+	}
+
+	fn matching_probe(&self, index: usize, device_info: DeviceInfo) -> MatchResult
+	{
+		// Consider the serial to match if it equals that of the device or if one was not specified at all.
+		let serial_matches = self.serial.as_deref().is_none_or(|s| Some(s) == device_info.serial_number());
+
+		// Consider the index to match if it equals that of the device or if one was not specified at all.
+		let index_matches = self.index.is_none_or(|needle| needle == index);
+
+		// Consider the port to match if it equals that of the device or if one was not specified at all.
+		let port_matches = self.port.as_ref().is_none_or(|p| {
+			let port = PortId::new(&device_info);
+
+			p == &port
+		});
+
+		// Finally, check the provided matchers.
+		if index_matches && port_matches && serial_matches {
+			match BmpDevice::from_usb_device(device_info) {
+				Ok(bmpdev) => MatchResult::Found(bmpdev),
+				Err(e) => MatchResult::Error(e),
+			}
+		} else {
+			MatchResult::NoMatch(device_info)
+		}
+	}
+}
+
+#[derive(Debug, Default)]
+pub struct BmpMatchResults
+{
+	pub found: Vec<BmpDevice>,
+	pub filtered_out: Vec<DeviceInfo>,
+	pub errors: Vec<Report>,
+}
+
+impl FromIterator<MatchResult> for BmpMatchResults
+{
+	fn from_iter<I: IntoIterator<Item = MatchResult>>(iter: I) -> Self
```

Comment:
Please make a note of what this function exists to do - it took a couple of read throughs to make sense of it and how it connects with `.collect()` above (not complaining about that, using the language's utilities like this is smart) so perhaps document that the point of this function is to implement the collect interface and turn a set of MatchResult into a fully populated BmpMatchResults piecewise.

This then also fills in the documentation gap it creates in the docs.rs docs.

Reply:
We're asking for the function (not the outer trait impl) to be documented which then fills the gap on why this exists and what it's solving. Set out what the function intends to do.

Reply:
It would be useful to include a line such as `/// This implements the internals of `.collect()` on an iterator to convert the `MatchResult`s into a `BmpMatchResults`